### PR TITLE
Implement pagination for MSC2666

### PR DESCRIPTION
**Pull request by tulir**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/rest/client/mutual_rooms.py [hunk 1]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
...
         user_id = user_ids[0]
 
         requester = await self.auth.get_user_by_req(request)
         if user_id == requester.user.to_string():
             raise SynapseError(
-                HTTPStatus.UNPROCESSABLE_ENTITY,
+                HTTPStatus.BAD_REQUEST,
                 "You cannot request a list of shared rooms with yourself",
-                errcode=Codes.INVALID_PARAM,
+                errcode=Codes.UNKNOWN,
```
Comment (tulir) [inline synapse/rest/client/mutual_rooms.py:102]: This change was made a couple years back on the MSC because 422 is a weird http status that nobody uses

File: synapse/rest/client/mutual_rooms.py [hunk 2]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
...
             )
 
-        rooms = await self.store.get_mutual_rooms_between_users(
-            frozenset((requester.user.to_string(), user_id))
+        rooms = sorted(
+            await self.store.get_mutual_rooms_between_users(
+                frozenset((requester.user.to_string(), user_id))
+            )
         )
```
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py]: Err, shall we sort in the database instead?

File: synapse/rest/client/mutual_rooms.py [hunk 3]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
...
+
+        if len(rooms) < MUTUAL_ROOMS_BATCH_LIMIT:
+            # We've reached the end of the list, don't return a batch token
+            return 200, {"joined": rooms}
+
+        rooms = rooms[:MUTUAL_ROOMS_BATCH_LIMIT]
+        # We use urlsafe unpadded base64 encoding for the batch token in order to
+        # handle funny room IDs in old pre-v12 rooms properly. We also truncate it
+        # to stay within the 255-character limit of opaque tokens.
+        next_batch = encode_base64(rooms[-1].encode("utf-8"), urlsafe=True)[:255]
```
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py:132]: We should acknowledge in a comment that this list can and will change underneath the user, and as such may omit or duplicate a room - but that otherwise this is a simple and pragmatic solution.

File: synapse/rest/client/mutual_rooms.py
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py]:
```suggestion
    GET /uk.half-shot.msc2666/user/mutual_rooms?user_id={user_id}&from={token} HTTP/1.1
```

File: synapse/rest/client/mutual_rooms.py [hunk 4]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
                 errcode=Codes.INVALID_PARAM,
             )
 
-        # We don't do batching, so a batch token is illegal by default
-        if b"batch_token" in args:
-            raise SynapseError(
-                HTTPStatus.BAD_REQUEST,
-                "Unknown batch_token",
```
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py:71]: Wow this was an exceptionally unhelpful error.

File: synapse/rest/client/mutual_rooms.py [hunk 5]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
...
-                HTTPStatus.UNPROCESSABLE_ENTITY,
+                HTTPStatus.BAD_REQUEST,
                 "You cannot request a list of shared rooms with yourself",
-                errcode=Codes.INVALID_PARAM,
+                errcode=Codes.UNKNOWN,
             )
 
-        rooms = await self.store.get_mutual_rooms_between_users(
-            frozenset((requester.user.to_string(), user_id))
+        rooms = sorted(
```
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py:107]:
```suggestion
        # Sort here instead of the database function, so that we don't expose
        # clients to any unrelated changes to the sorting algorithm.
        rooms = sorted(
```

File: tests/rest/client/test_mutual_rooms.py
Comment (anoadragon453) [inline tests/rest/client/test_mutual_rooms.py:1]: If you're feeling up to it, I would love to see some [Complement](https://github.com/matrix-org/complement) tests for this endpoint as well.

But not a blocker for this PR (or the spec).

File: tests/rest/client/test_mutual_rooms.py [hunk 6]
```diff
@@ -141,6 +145,30 @@ def _check_mutual_rooms_with(
         for room_id_id in channel.json_body["joined"]:
             self.assertIn(room_id_id, [room_id_one, room_id_two])
 
+    def test_shared_room_list_pagination(self) -> None:
+        u1 = self.register_user("user1", "pass")
+        u1_token = self.login(u1, "pass")
+        u2 = self.register_user("user2", "pass")
+        u2_token = self.login(u2, "pass")
+        self.get_success(self.hs.get_datastores().main.set_ratelimit_for_user(u1, 0, 0))
```
Comment (anoadragon453) [inline tests/rest/client/test_mutual_rooms.py]: We should be eliminating the ratelimit in the test homeserver config already: https://github.com/element-hq/synapse/blob/a096fba969f43fb89d946d14a8b41ed23b7538f6/tests/utils.py#L186-L203

Did you find this was necessary? (Or is it just due to us missing `rc_room_creation` in that config?)

File: synapse/rest/client/mutual_rooms.py [hunk 7]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
...
+            await self.store.get_mutual_rooms_between_users(
+                frozenset((requester.user.to_string(), user_id))
+            )
         )
 
-        return 200, {"joined": list(rooms)}
+        if from_batch:
+            # A from_batch token was provided, so cut off any rooms where the ID is
+            # lower than or equal to the token
+            rooms = rooms[bisect(rooms, from_batch) :]
```
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py:122]: Note that in the case where an invalid `from` token is supplied (but is still a base64-decodable string), `bisect` will still return an index where it can be inserted and the user will get a list.

I think this is fine, since it's unlikely that someone will provide a valid base64 string that's not the `from` token (and thus we don't really need to handle this edge case with a specific error). But we should probably leave a comment about it.

File: tests/rest/client/test_mutual_rooms.py [hunk 8]
```diff
@@ -141,6 +145,30 @@ def _check_mutual_rooms_with(
...
+        channel = self._get_mutual_rooms(u1_token, u2)
+        self.assertEqual(200, channel.code, channel.result)
+        self.assertEqual(channel.json_body["joined"], room_ids[0:10])
+        self.assertIn("next_batch", channel.json_body)
+
+        channel = self._get_mutual_rooms(u1_token, u2, channel.json_body["next_batch"])
+        self.assertEqual(200, channel.code, channel.result)
+        self.assertEqual(channel.json_body["joined"], room_ids[10:20])
+        self.assertNotIn("next_batch", channel.json_body)
+
```
Comment (anoadragon453) [inline tests/rest/client/test_mutual_rooms.py:176]: Could you add some additional tests for:

- an invalid from token being passed (specifically something that would trigger INVALID_PARAM).
- a batch limit of 10, and a set of 10 mutual rooms. check that next_batch is not provided.
- trying to get mutual rooms with a user that does not exist.

File: synapse/rest/client/mutual_rooms.py [hunk 9]
```diff
@@ -64,29 +92,44 @@ async def on_GET(self, request: SynapseRequest) -> tuple[int, JsonDict]:
...
+
+        if len(rooms) < MUTUAL_ROOMS_BATCH_LIMIT:
+            # We've reached the end of the list, don't return a batch token
+            return 200, {"joined": rooms}
```
Comment (anoadragon453) [inline synapse/rest/client/mutual_rooms.py]: If we're returning a list of exactly 100 items, then the user will still get back a `next_batch` token, no? But their next request will be unnecessary.


```suggestion
        if len(rooms) <= MUTUAL_ROOMS_BATCH_LIMIT:
            # We've reached the end of the list, don't return a batch token
            return 200, {"joined": rooms}
```

**Review by anoadragon453 — CHANGES_REQUESTED** [review]
A few things here and there, but on the whole this is mostly good out of the gate. Thank you for moving this highly-requested feature along @tulir!

File: tests/rest/client/test_mutual_rooms.py (hunk 6 above)
Comment (tulir) [reply tests/rest/client/test_mutual_rooms.py]: The test exploded with a 429 in room creation until I added that, I guess it's the missing `rc_room_creation`

File: synapse/rest/client/mutual_rooms.py (hunk 7 above)
Comment (tulir) [reply synapse/rest/client/mutual_rooms.py:122]: Yeah it's intentional that the exact value of the token isn't validated. Due to the truncation it's not even guaranteed to be a real room ID, plus even if it is, the room might have disappeared after the first request.

File: tests/rest/client/test_mutual_rooms.py (hunk 6 above)
Comment (tulir) [reply tests/rest/client/test_mutual_rooms.py]: I had to add both `rc_room_creation` and `rc_invites`->`per_issuer` to make it pass without the ratelimit override

File: synapse/rest/client/mutual_rooms.py (hunk 3 above)
Comment (tulir) [reply synapse/rest/client/mutual_rooms.py:132]: I updated the comment in the bisect code

File: tests/rest/client/test_mutual_rooms.py (hunk 8 above)
Comment (tulir) [reply tests/rest/client/test_mutual_rooms.py:176]: Done

**Comment by anoadragon453** [comment]
CI failures are the result of flaky tests.
